share/eds/store.go

```diff
@@ -231,12 +264,30 @@ func (s *Store) getAccessor(ctx context.Context, key shard.Key) (*accessorWithBl
 		if res.Error != nil {
 			return nil, fmt.Errorf("failed to acquire shard: %w", res.Error)
 		}
-		return s.cache.Add(key, res.Accessor)
+		return res.Accessor, nil
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	}
 }
 
+func (s *Store) getAccessor(ctx context.Context, key shard.Key) (*accessorWithBlockstore, error) {
+	// try to fetch from cache
+	accessor, err := s.cache.Get(key)
+	if err != nil && err != errCacheMiss {
+		log.Errorw("unexpected error while reading key from bs cache %s: %s", key, err)
+	}
+	if accessor != nil {
+		return accessor, nil
+	}
+
+	// wasn't found in cache, so acquire it and add to cache
+	shardAccessor, err := s.acquireShard(ctx, key)
+	if err != nil {
+		return nil, err
+	}
+	return s.cache.Add(key, shardAccessor)
```

Comment:
Does the acquireShard has internal synchronisation inside dagstore? If 2 routines simultaneously do not find same key in the cache and then try to acquire it and save it, could that lead to any problems?

Reply:
No, no, there shouldn't be any problems; the DagStore is synchronized and designed around thread safety